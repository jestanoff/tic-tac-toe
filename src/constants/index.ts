export const SYMBOLS = {
  1: 'cross',
  2: 'circle',
};

export const CSS_CLASSES = {
  0: ['top', 'left'],
  1: ['top', 'center'],
  2: ['top', 'right'],
  3: ['left', 'middle'],
  4: ['center', 'middle'],
  5: ['right', 'middle'],
  6: ['bottom', 'left'],
  7: ['bottom', 'center'],
  8: ['bottom', 'right'],
};

interface Coords {
  0: Array<number>;
  1: Array<number>;
  2: Array<number>;
  3: Array<number>;
  4: Array<number>;
  5: Array<number>;
  6: Array<number>;
  7: Array<number>;
}

export const COORDS: Coords = {
  //  x1, y1, x2, y2
  0: [30, 82, 470, 82], // horizontal
  1: [30, 250, 470, 250],
  2: [30, 418, 470, 418],
  3: [82, 30, 82, 470], // vertical
  4: [250, 30, 250, 470],
  5: [418, 30, 418, 470],
  6: [40, 40, 460, 460], // diagonal
  7: [460, 40, 40, 460],
};

export const AI_WAITING_TIME = 700;
export const NUMBER_OF_SPACES = 9;
export const PLAYER_X = 1;
export const PLAYER_O = 2;
export const X_WON = 1;
export const O_WON = 2;
export const DRAW = 0;
export const UNRESOLVED = -1;
export const M_DASH = '\u2014';
export const EASY = 'Easy';
export const IMPOSSIBLE = 'Impossible';
export const RESET = 'Reset';
export const DARK_GRAY = '#444444';
export const WHITE = '#FFFFFF';
