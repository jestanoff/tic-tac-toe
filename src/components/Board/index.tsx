import * as React from 'react';
import * as PropTypes from 'prop-types';
import Space from 'components/Space';
import styles from './Board.css';
import { CSS_CLASSES, NUM_OF_CELLS } from '../../constants';

const Board = ({ boardStatus, handleSpaceClick }) => (
  <section className={styles.container}>
    {Array.from({ length: NUM_OF_CELLS }, (_, i) => i).map((index) => (
      <Space
        cssClasses={CSS_CLASSES[index]}
        handleSpaceClick={handleSpaceClick}
        id={index}
        key={index}
        status={boardStatus[index]}
      />
    ))}
  </section>
);

Board.propTypes = {
  boardStatus: PropTypes.arrayOf(PropTypes.number).isRequired,
  handleSpaceClick: PropTypes.func.isRequired,
};

export default React.memo(Board);
