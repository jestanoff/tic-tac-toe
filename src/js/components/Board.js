import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Cell from './Cell';
import styles from '../../css/board.css';
import { CSS_CLASSES, NUM_OF_CELLS } from '../constants/index';

const Board = ({ boardStatus, handleCellClick }) => (
  <section className={styles.container}>
    {Array.from({ length: NUM_OF_CELLS }, (_, i) => i).map((index) => (
      <Cell
        cssClasses={CSS_CLASSES[index]}
        handleCellClick={handleCellClick}
        id={index}
        key={index}
        status={boardStatus[index]}
      />
    ))}
  </section>
);

Board.propTypes = {
  boardStatus: PropTypes.arrayOf(PropTypes.number).isRequired,
  handleCellClick: PropTypes.func.isRequired,
};

export default memo(Board);
