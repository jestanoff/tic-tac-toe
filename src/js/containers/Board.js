import React, { PropTypes } from 'react';
import Cell from '../components/Cell';
import styles from '../../css/board.css';
import { CSS_CLASSES, NUM_OF_CELLS } from '../constants/constants';

const Board = ({ boardStatus, handleCellClick }) => (
    <section className={ styles.container }>
        { Array.from({ length: NUM_OF_CELLS }, (_, i) => i).map(index => (
            <Cell
              cssClasses={ CSS_CLASSES[index] }
              handleCellClick={ handleCellClick }
              id={ index }
              key={ index }
              status={ boardStatus[index] }
            />
        ))}
    </section>
);

Board.propTypes = {
    boardStatus: PropTypes.arrayOf(PropTypes.number).isRequired,
    handleCellClick: PropTypes.func.isRequired,
};

export default Board;
