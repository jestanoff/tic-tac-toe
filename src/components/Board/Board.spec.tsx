import * as React from 'react';
import { render, screen } from '@testing-library/react';
import Board from './index';
import { NUMBER_OF_SPACES } from '../../constants';

describe('<Board />', () => {
  const boardStatus = [0, 2, 1, 0, 0, 0, 1, 0, 2];
  const handleCellClick = () => {};

  test('should render proper markup', () => {
    const { asFragment } = render(<Board boardStatus={boardStatus} handleSpaceClick={handleCellClick} />);

    expect(asFragment()).toMatchSnapshot();
  })

  test('should render the board according to its status', () => {
    render(<Board boardStatus={boardStatus} handleSpaceClick={handleCellClick} />);

    expect(screen.getAllByRole('button')).toHaveLength(NUMBER_OF_SPACES);
    expect(screen.getByRole('button', { name: /Top left empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Top center space marked by circle/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Top right space marked by cross/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Left middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Center middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Right middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Bottom left space marked by cross/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Bottom center empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Bottom right space marked by circle/i })).toBeInTheDocument();
  });
});
