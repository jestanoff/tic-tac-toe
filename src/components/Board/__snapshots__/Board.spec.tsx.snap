// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> should render proper markup 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <button
      class="space top left"
      title="Top left empty space"
    />
    <button
      class="space top center"
      title="Top center space marked by circle"
    >
      <div
        class="container"
      >
        <svg
          class="svg"
          data-testid="circle-mark"
          preserveAspectRatio="xMinYMin meet"
          version="1.1"
          viewBox="0 0 500 500"
        >
          <circle
            cx="250"
            cy="250"
            fill="transparent"
            r="125"
            stroke="#FFFFFF"
            stroke-dasharray="800"
            stroke-dashoffset="-800"
            stroke-width="45"
          >
            <animate
              attributeName="stroke-dashoffset"
              attributeType="XML"
              dur="250ms"
              fill="freeze"
              from="-800"
              repeatCount="1"
              to="0"
            />
          </circle>
        </svg>
      </div>
    </button>
    <button
      class="space top right"
      title="Top right space marked by cross"
    >
      <div
        class="container"
      >
        <svg
          class="svg"
          data-testid="cross-mark"
          preserveAspectRatio="xMinYMin meet"
          version="1.1"
          viewBox="0 0 500 500"
        >
          <line
            stroke="#444444"
            stroke-width="50"
            x1="130"
            x2="130"
            y1="130"
            y2="130"
          >
            <animate
              attributeName="x2"
              attributeType="XML"
              dur="125ms"
              fill="freeze"
              from="130"
              id="first"
              repeatCount="1"
              to="370"
            />
            <animate
              attributeName="y2"
              attributeType="XML"
              dur="125ms"
              fill="freeze"
              from="130"
              repeatCount="1"
              to="370"
            />
          </line>
          <line
            opacity="0"
            stroke="#444444"
            stroke-width="50"
            x1="370"
            x2="370"
            y1="130"
            y2="130"
          >
            <animate
              attributeName="opacity"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="0"
              repeatCount="1"
              to="1"
            />
            <animate
              attributeName="x2"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="370"
              repeatCount="1"
              to="130"
            />
            <animate
              attributeName="y2"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="130"
              repeatCount="1"
              to="370"
            />
          </line>
        </svg>
      </div>
    </button>
    <button
      class="space left middle"
      title="Left middle empty space"
    />
    <button
      class="space center middle"
      title="Center middle empty space"
    />
    <button
      class="space right middle"
      title="Right middle empty space"
    />
    <button
      class="space bottom left"
      title="Bottom left space marked by cross"
    >
      <div
        class="container"
      >
        <svg
          class="svg"
          data-testid="cross-mark"
          preserveAspectRatio="xMinYMin meet"
          version="1.1"
          viewBox="0 0 500 500"
        >
          <line
            stroke="#444444"
            stroke-width="50"
            x1="130"
            x2="130"
            y1="130"
            y2="130"
          >
            <animate
              attributeName="x2"
              attributeType="XML"
              dur="125ms"
              fill="freeze"
              from="130"
              id="first"
              repeatCount="1"
              to="370"
            />
            <animate
              attributeName="y2"
              attributeType="XML"
              dur="125ms"
              fill="freeze"
              from="130"
              repeatCount="1"
              to="370"
            />
          </line>
          <line
            opacity="0"
            stroke="#444444"
            stroke-width="50"
            x1="370"
            x2="370"
            y1="130"
            y2="130"
          >
            <animate
              attributeName="opacity"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="0"
              repeatCount="1"
              to="1"
            />
            <animate
              attributeName="x2"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="370"
              repeatCount="1"
              to="130"
            />
            <animate
              attributeName="y2"
              attributeType="XML"
              begin="125ms"
              dur="125ms"
              fill="freeze"
              from="130"
              repeatCount="1"
              to="370"
            />
          </line>
        </svg>
      </div>
    </button>
    <button
      class="space bottom center"
      title="Bottom center empty space"
    />
    <button
      class="space bottom right"
      title="Bottom right space marked by circle"
    >
      <div
        class="container"
      >
        <svg
          class="svg"
          data-testid="circle-mark"
          preserveAspectRatio="xMinYMin meet"
          version="1.1"
          viewBox="0 0 500 500"
        >
          <circle
            cx="250"
            cy="250"
            fill="transparent"
            r="125"
            stroke="#FFFFFF"
            stroke-dasharray="800"
            stroke-dashoffset="-800"
            stroke-width="45"
          >
            <animate
              attributeName="stroke-dashoffset"
              attributeType="XML"
              dur="250ms"
              fill="freeze"
              from="-800"
              repeatCount="1"
              to="0"
            />
          </circle>
        </svg>
      </div>
    </button>
  </section>
</DocumentFragment>
`;
